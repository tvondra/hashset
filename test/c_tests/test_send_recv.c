#include <stdio.h>
#include <stdlib.h>
#include <libpq-fe.h>

void exit_nicely(PGconn *conn) {
	PQfinish(conn);
	exit(1);
}

int main() {
	/* Connect to database specified by the PGDATABASE environment variable */
	const char	*conninfo = "host=localhost port=5432";
	PGconn		*conn = PQconnectdb(conninfo);
	if (PQstatus(conn) != CONNECTION_OK) {
		fprintf(stderr, "Connection to database failed: %s", PQerrorMessage(conn));
		exit_nicely(conn);
	}

	/* Create extension */
	PQexec(conn, "CREATE EXTENSION IF NOT EXISTS hashset");

	/* Create temporary table */
	PQexec(conn, "CREATE TABLE IF NOT EXISTS test_hashset_send_recv (hashset_col hashset)");

	/* Enable binary output */
	PQexec(conn, "SET bytea_output = 'escape'");

	/* Insert dummy data */
	const char *insert_command = "INSERT INTO test_hashset_send_recv (hashset_col) VALUES ('{1,2,3}'::hashset)";
	PGresult *res = PQexec(conn, insert_command);
	if (PQresultStatus(res) != PGRES_COMMAND_OK) {
		fprintf(stderr, "INSERT failed: %s", PQerrorMessage(conn));
		PQclear(res);
		exit_nicely(conn);
	}
	PQclear(res);

	/* Fetch the data in binary format */
	const char *select_command = "SELECT hashset_col FROM test_hashset_send_recv";
	int resultFormat = 1;  /* 0 = text, 1 = binary */
	res = PQexecParams(conn, select_command, 0, NULL, NULL, NULL, NULL, resultFormat);
	if (PQresultStatus(res) != PGRES_TUPLES_OK) {
		fprintf(stderr, "SELECT failed: %s", PQerrorMessage(conn));
		PQclear(res);
		exit_nicely(conn);
	}

	/* Store binary data for later use */
	const char *binary_data = PQgetvalue(res, 0, 0);
	int binary_data_length = PQgetlength(res, 0, 0);
	PQclear(res);

	/* Re-insert the binary data */
	const char *insert_binary_command = "INSERT INTO test_hashset_send_recv (hashset_col) VALUES ($1)";
	const char *paramValues[1] = {binary_data};
	int paramLengths[1] = {binary_data_length};
	int paramFormats[1] = {1}; /* binary format */
	res = PQexecParams(conn, insert_binary_command, 1, NULL, paramValues, paramLengths, paramFormats, 0);
	if (PQresultStatus(res) != PGRES_COMMAND_OK) {
		fprintf(stderr, "INSERT failed: %s", PQerrorMessage(conn));
		PQclear(res);
		exit_nicely(conn);
	}
	PQclear(res);

	/* Check the data */
	const char *check_command = "SELECT COUNT(DISTINCT hashset_col::text) AS unique_count, COUNT(*) FROM test_hashset_send_recv";
	res = PQexec(conn, check_command);
	if (PQresultStatus(res) != PGRES_TUPLES_OK) {
		fprintf(stderr, "SELECT failed: %s", PQerrorMessage(conn));
		PQclear(res);
		exit_nicely(conn);
	}

	/* Print the results */
	printf("unique_count: %s\n", PQgetvalue(res, 0, 0));
	printf("count: %s\n", PQgetvalue(res, 0, 1));
	PQclear(res);

	/* Disconnect */
	PQfinish(conn);

	return 0;
}
