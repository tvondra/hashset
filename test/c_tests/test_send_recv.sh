#!/bin/sh

# Get the directory of this script
SCRIPT_DIR="$(dirname "$(realpath "$0")")"

# Set up database
export PGDATABASE=test_hashset_send_recv
dropdb --if-exists "$PGDATABASE"
createdb

# Define directories
EXPECTED_DIR="$SCRIPT_DIR/../expected"
RESULTS_DIR="$SCRIPT_DIR/../results"

# Create the results directory if it doesn't exist
mkdir -p "$RESULTS_DIR"

# Run the C test and save its output to the results directory
"$SCRIPT_DIR/test_send_recv" > "$RESULTS_DIR/test_send_recv.out"

printf "test test_send_recv               ... "

# Compare the actual output with the expected output
if diff -q "$RESULTS_DIR/test_send_recv.out" "$EXPECTED_DIR/test_send_recv.out" > /dev/null 2>&1; then
    echo "ok"
    # Clean up by removing the results directory if the test passed
    rm -r "$RESULTS_DIR"
else
    echo "failed"
    git diff --no-index --color "$EXPECTED_DIR/test_send_recv.out" "$RESULTS_DIR/test_send_recv.out"
fi
