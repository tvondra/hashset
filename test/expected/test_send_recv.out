unique_count: 1
count: 2
